src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+    d[q] = (q - v[k]) * (q - v[k]) + f[v[k]];
+  }
+  return d;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+DistanceTransform2D(MatType &im, const double inf)
+{
+  arma::vec f(std::max(im.n_rows, im.n_cols));
+  // transform along columns
+  for (size_t x = 0; x < im.n_cols; ++x)
+  {
+    f.subvec(0, im.n_rows - 1) = im.col(x);
+    arma::vec d = this->DistanceTransform1D(f, im.n_rows, inf);
+    im.col(x) = d;
+  }
+
+  // transform along rows
+  for (size_t y = 0; y < im.n_rows; y++)
+  {
+    f.subvec(0, im.n_cols - 1) = im.row(y).t();
+    arma::vec d = this->DistanceTransform1D(f, im.n_cols, inf);
+    im.row(y) = d.t();
+  }
+}
+
+/* euclidean distance transform of binary image using squared distance */
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+DistanceTransformImage(MatType const &im, double on)
+{
+  //need a large value but not infinity.
+  double inf = 999999.99;
+  MatType out = MatType(im.n_rows, im.n_cols, arma::fill::zeros);
+  out.elem( find(im != on) ).fill(inf);
+  this->DistanceTransform2D(out, inf);
+  return out;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+CopyMakeBorder(CubeType const &InImage, size_t top, 
+               size_t left, size_t bottom, size_t right)
+{
+  CubeType OutImage(InImage.n_rows + top + bottom, InImage.n_cols + left + right, InImage.n_slices);
+
+  for(size_t i = 0; i < InImage.n_slices; ++i)
+  {
+    OutImage.slice(i).submat(top, left, InImage.n_rows + top - 1, InImage.n_cols + left - 1)
+     = InImage.slice(i);
+    
+    for(size_t j = 0; j < right; ++j)
+    {
+      OutImage.slice(i).col(InImage.n_cols + left + j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(InImage.n_cols - j - 1);  
+    }
+
+    for(size_t j = 0; j < left; ++j)
+    {
+      OutImage.slice(i).col(j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(left - 1 - j);
+    }
+
+    for(size_t j = 0; j < top; j++)
+    {
+
+      OutImage.slice(i).row(j)
+      = OutImage.slice(i).row(2 * top - 1 - j);
+    }
+     
+    for(size_t j = 0; j < bottom; j++)
+    {
+      OutImage.slice(i).row(InImage.n_rows + top + j)
+      = OutImage.slice(i).row(InImage.n_rows + top - j - 1);
+    }
+
+  }
+  return OutImage;  
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+RGB2LUV(CubeType const &InImage)
+{
+  //assert type is double or float.
+  double a, y0, maxi;
+  a = std::pow(29.0, 3) / 27.0;
+  y0 = 8.0 / a;
+  maxi = 1.0 / 270.0;
+
+  arma::vec table(1064);
+  for (size_t i = 0; i <= 1024; ++i)
+  {
+    table(i) = i / 1024.0;
+    
+    if (table(i) > y0)
+      table(i) = 116 * pow(table(i), 1.0/3.0) - 16.0;
+    else
+      table(i) = table(i) * a;
+
+    table(i) = table(i) * maxi;
+  }
+  for(size_t i = 1025; i < table.n_elem; ++i)
+  {
+    table(i) = table(i - 1);
+  }
+
```

Comment:
I think it is a good idea, to make it run faster we could try to vectorize the codes more, I remember I post a link of matlab before([click me](http://pages.cs.wisc.edu/~dyer/cs766/hw/hw2/code/rgb2luv.m))